File: AppFeedback/ScreenVideoCaptureSession.m
//
//  ScreenVideoCaptureSession.m
//  ScreenVideoCaptureSession
//
//  Copyright (c) 2018 Yahoo Japan Corporation.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.


#import "ScreenVideoCaptureSession.h"
#import "ScreenCapture.h"
#import "Logger.h"

@import AVFoundation;
@import UIKit;

@interface ScreenVideoCaptureSession ()

@property (atomic) BOOL recording;
@property (nonatomic, strong) NSDate *startedAt;
@property (nonatomic, strong) NSURL *outputFile;
@property (nonatomic) UIInterfaceOrientation initOrientation;
@property (nonatomic) CGSize initScreenSize;
@property (nonatomic) CGFloat initScreenScale;
@property (nonatomic, copy) StopRecordingCallBack callback;
@property (nonatomic, assign) float frameRate;
@property (nonatomic, strong) AVAssetWriter *videoWriter;
@property (nonatomic, strong) AVAssetWriterInput *videoWriterInput;
@property (nonatomic, strong) AVAssetWriterInputPixelBufferAdaptor *avAdaptor;
@property (nonatomic, strong) NSTimer *timer;

@end


@implementation ScreenVideoCaptureSession

- (id) init {
    self = [super init];
    if (self) {
        self.frameRate = 20.0f; // frames per seconds
    }
    return self;
}

- (float)interval {
    return 1.0 / self.frameRate;
}

- (void)cleanupWriter {
    _avAdaptor = nil;
    _videoWriterInput = nil;
    _videoWriter = nil;
}

- (void)dealloc {
    [self cleanupWriter];

    _outputFile = nil;
    _startedAt = nil;
}

- (NSURL*)tempFileURL:(NSString *)name {
    NSString* outputPath = [NSString pathWithComponents:@[NSTemporaryDirectory(), name]];
    return [[NSURL alloc] initFileURLWithPath:outputPath];
}

- (void)clearFile:(NSURL *)path {
    NSFileManager* fileManager = [NSFileManager defaultManager];

    NSError* error;

    if ([path checkResourceIsReachableAndReturnError:&error]) {
        if (![fileManager removeItemAtURL:path error:&error]) {
            Log(@"Could not delete old recording file at path:  %@", path.absoluteString);
        }
    }
}

-(BOOL) setUpWriter {
    self.outputFile = [self tempFileURL:@"app_feedback_video.mp4"];
    [self clearFile:self.outputFile];

    NSError* error = nil;
    self.videoWriter = [[AVAssetWriter alloc] initWithURL:self.outputFile
                                                 fileType:AVFileTypeQuickTimeMovie
                                                    error:&error];
    NSParameterAssert(self.videoWriter);

    //Configure video
    NSDictionary* videoCompressionProps = @{
                                            // 明示的に設定しなくて良さそう
                                            //AVVideoAverageBitRateKey: @(1024.0*1024.0),
                                            //AVVideoMaxKeyFrameIntervalKey: @(self.frameRate),
                                            };

    CGSize imageSize = self.initScreenSize;
    CGFloat scale = self.initScreenScale;

    imageSize.height *= scale;
    imageSize.width *= scale;

    NSDictionary* videoSettings = @{
                                    AVVideoCodecKey: AVVideoCodecH264,
                                    AVVideoWidthKey: @(imageSize.width),
                                    AVVideoHeightKey: @(imageSize.height),
                                    AVVideoCompressionPropertiesKey: videoCompressionProps,
                                    };
    
    self.videoWriterInput = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeVideo
                                                          outputSettings:videoSettings];

    NSParameterAssert(self.videoWriterInput);
    self.videoWriterInput.expectsMediaDataInRealTime = YES;
    NSDictionary* bufferAttributes = [NSDictionary dictionaryWithObjectsAndKeys:
                                      @(kCVPixelFormatType_32BGRA), kCVPixelBufferPixelFormatTypeKey,
                                      nil];
    
    self.avAdaptor = [AVAssetWriterInputPixelBufferAdaptor
                      assetWriterInputPixelBufferAdaptorWithAssetWriterInput:self.videoWriterInput
                      sourcePixelBufferAttributes:bufferAttributes];

    //add input
    [self.videoWriter addInput:self.videoWriterInput];
    [self.videoWriter startWriting];
    [self.videoWriter startSessionAtSourceTime:CMTimeMake(0, 1000)];

    return YES;
}

- (CGFloat)fixedImageScale {
    CGSize size = UIScreen.mainScreen.bounds.size;
    CGFloat scale = UIScreen.mainScreen.scale;
    if (size.width == 667 && size.height == 375 && scale == 2.0) {
        // iPhone 6, 7 などの4.7 インチ端末の横向きでなぜか動画が正常にキャプチャできないため
        // scale をいじる。この scale だとなぜかうまくいく (^^;
        return 1.5;
    } else if (size.width == 375 && size.height == 812 && scale == 3.0) {
        // iPhone X ポートレートモードでキャプチャを正常に取れるようにする対応
        return 2.0;
    } else if (size.width == 812 && size.height == 375 && scale == 3.0) {
        // iPhone X ランドスケープモードでキャプチャを正常に取れるようにする対応
        return 1.2;
    }
    return scale;
}

- (void)runRecordFrameLoop:(float)delay {
    if (!self.recording) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            [self completeRecordingSession];
        });
        return;
    }

    dispatch_queue_t q = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, delay * NSEC_PER_SEC), q, ^{
        [self recordFrame];
    });
}

- (void) recordFrame {
    NSDate* start = [NSDate date];

    __block UIImage *image = nil;
    dispatch_sync(dispatch_get_main_queue(), ^{
        image = [ScreenCapture captureImageWithScale:self.initScreenScale
                                                size:self.initScreenSize
                                     baseOrientation:self.initOrientation];
    });

    float millisElapsed = [[NSDate date] timeIntervalSinceDate:self.startedAt] * 1000.0;
    [self writeVideoFrameAtTime:CMTimeMake((int)millisElapsed, 1000) image:image];

    float processingSeconds = [[NSDate date] timeIntervalSinceDate:start];
    float delayRemaining = [self interval] - processingSeconds;
    if (delayRemaining <= 0.0) {
        // Log(@"delayRemaining %.2f", delayRemaining);
    }

    float delay = delayRemaining > 0.0 ? delayRemaining : [self interval];
    [self runRecordFrameLoop:delay];
}

- (void) completeRecordingSession {
    // このメソッドはバックグラウンドで呼ばれる

    [self.videoWriterInput markAsFinished];

    // Wait for the video
    int status = self.videoWriter.status;
    while (status == AVAssetWriterStatusUnknown) {
        // Log(@"Waiting...");
        [NSThread sleepForTimeInterval:0.5f];
        status = self.videoWriter.status;
    }

    [self.videoWriter finishWritingWithCompletionHandler:^{
        AVAssetWriterStatus status = self.videoWriter.status;
        
        if (self.videoWriter.error) {
            NSError *error = self.videoWriter.error;
            dispatch_async(dispatch_get_main_queue(), ^{
                self.callback(nil, error);
                self.callback = nil;
            });
        }
        else if (status == AVAssetWriterStatusCompleted) {
            dispatch_async(dispatch_get_main_queue(), ^{
                self.callback(self.outputFile, nil);
                self.callback = nil;
            });
        }
        [self cleanupWriter];
    }];
}

- (bool)startRecordingUntil:(float)timeLimitSecs callback:(StopRecordingCallBack)callback {
    if (self.recording) {
        Log(@"Already recording.");
        return NO;
    }

    self.initOrientation = UIApplication.sharedApplication.statusBarOrientation;
    self.initScreenSize = UIScreen.mainScreen.bounds.size;;
    self.initScreenScale = [self fixedImageScale];

    if (![self setUpWriter]) {
        Log(@"setUpWriter failed.");
        return NO;
    }
    
    self.callback = callback;
    self.startedAt = [NSDate date];
    self.timer = [NSTimer scheduledTimerWithTimeInterval:timeLimitSecs
                                                  target:self
                                                selector:@selector(timerFired:)
                                                userInfo:nil
                                                 repeats:NO];
    self.recording = true;
    [self runRecordFrameLoop:0];

    return YES;
}

- (void)stopRecording {
    if (self.recording) {
        [self.timer invalidate];
        self.timer = nil;
        self.recording = false;
    }
}

- (void)timerFired:(NSTimer *)timer {
    [self stopRecording];
}

-(void) writeVideoFrameAtTime:(CMTime)time image:(UIImage *)image {
    if (![self.videoWriterInput isReadyForMoreMediaData]) {
        Log(@"videoWriterInput not ready for video data");
        return;
    }
    
    if (!self.avAdaptor.pixelBufferPool) {
        Log(@"Error pixelBufferPool is nil.");
        return;
    }

    CVPixelBufferRef pixelBuffer = NULL;
    int status = CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, self.avAdaptor.pixelBufferPool, &pixelBuffer);
    if(status != 0){
        //could not get a buffer from the pool
        Log(@"Error creating pixel buffer:  status=%d", status);
        return;
    }

    CGImageRef cgImage = CGImageCreateCopy([image CGImage]);
    CFDataRef data = CGDataProviderCopyData(CGImageGetDataProvider(cgImage));
    
    // set image data into pixel buffer
    CVPixelBufferLockBaseAddress( pixelBuffer, 0 );
    uint8_t* destPixels = CVPixelBufferGetBaseAddress(pixelBuffer);
    CFRange range = CFRangeMake(0, CFDataGetLength(data));
    CFDataGetBytes(data, range, destPixels);
    
    BOOL appendSuccess = [self.avAdaptor appendPixelBuffer:pixelBuffer withPresentationTime:time];

    CVPixelBufferUnlockBaseAddress(pixelBuffer, 0);
    CVPixelBufferRelease(pixelBuffer);
    CFRelease(data);
    CGImageRelease(cgImage);

    if (!appendSuccess) {
        Log(@"appendPixelBuffer failed");
    }
}

@end

File: AppFeedback/ScreenVideoCaptureSession.swift
//
//  ScreenVideoCaptureSession.swift
//  ScreenVideoCaptureSession
//
//  Copyright (c) 2018 Yahoo Japan Corporation.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.


import Foundation
import AVKit

@objcMembers
class ScreenVideoCaptureSession: NSObject {
    typealias StopRecordingCallBack = (URL?, Error?) -> Void
    
    private let frameRate: Float = 20

    private(set) var recording: Bool = false
    
    private var startedAt: Date?
    private var outputFile: URL?
    private var initOrientation: UIInterfaceOrientation?
    private var initScreenSize: CGSize = .zero
    private var initScreenScale: CGFloat = 0
    private var callback: StopRecordingCallBack?
    private var videoWriter: AVAssetWriter?
    private var videoWriterInput: AVAssetWriterInput?
    private var avAdaptor: AVAssetWriterInputPixelBufferAdaptor?
    private var timer: Timer?
    
    override init() {
        super.init()
    }
    
    private var interval: Float {
        return 1 / frameRate
    }
    
    private func cleanupWriter() {
        avAdaptor = nil
        videoWriterInput = nil
        videoWriter = nil
    }
    
    deinit {
        cleanupWriter()
        outputFile = nil
        startedAt = nil
    }
    
    private func tempFileURL(_ name: String) -> URL? {
        let outputPath = NSString.path(withComponents: [NSTemporaryDirectory(), name])
        return URL(fileURLWithPath: outputPath)
    }
    
    private func clearFile(path: URL) {
        let fileManager = FileManager.default
        do {
            if try path.checkResourceIsReachable() {
                try fileManager.removeItem(at: path)
            }
        } catch {
            print("Could not delete old recording file at path: \(path.absoluteString)")
        }
    }
    
    private func setupWriter() -> Bool {
        guard let outputFile = tempFileURL("app_feedback_video.mp4") else {
            return false
        }
        
        clearFile(path: outputFile)
        self.outputFile = outputFile
        
        do {
            videoWriter = try AVAssetWriter(outputURL: outputFile, fileType: .mov)
            var imageSize = initScreenSize
            let scale = initScreenScale
            
            imageSize.height *= scale
            imageSize.width *= scale
            
            let videSettings: [String: Any] = [
                AVVideoCodecKey: AVVideoCodecH264,
                AVVideoWidthKey: imageSize.width,
                AVVideoHeightKey: imageSize.height,
                AVVideoCompressionPropertiesKey: [String: Any]()
            ]
            
            let videoWriterInput = AVAssetWriterInput(mediaType: .video, outputSettings: videSettings)
            videoWriterInput.expectsMediaDataInRealTime = true
            let bufferAttributes: [String: Any] = [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]
            avAdaptor = AVAssetWriterInputPixelBufferAdaptor(assetWriterInput: videoWriterInput, sourcePixelBufferAttributes: bufferAttributes)
            self.videoWriterInput = videoWriterInput
            
            
            videoWriter?.add(videoWriterInput)
            videoWriter?.startWriting()
            videoWriter?.startSession(atSourceTime: CMTime(value: 0, timescale: 1000))
        } catch {
        }
        
        return true
    }
    
    var fixedImageScale: CGFloat {
        let size = UIScreen.main.bounds.size
        let scale = UIScreen.main.scale
        
        if size.width == 667 && size.height == 375 && scale == 2.0 {
            return 1.5
        } else if size.width == 375 && size.height == 812 && scale == 3.0  {
            return 2.0
        } else if size.width == 812 && size.height == 375 && scale == 3.0 {
            return 1.2
        } else {
            return scale
        }
    }
    
    func runRecordFrameLoop(delay: TimeInterval) {
        if recording {
            DispatchQueue.global().async { [weak self] in
                self?.completeRecordingSession()
            }
        }
        
        DispatchQueue.global().asyncAfter(deadline: .now() + delay) { [weak self] in
            self?.recordFrame()
        }
    }
    
    func recordFrame() {
        let start = Date()
        var image: UIImage?
        DispatchQueue.main.sync {
            // TODO
        }
        
        if let startedAt = startedAt, let image = image {
            let millisElapsed = Date().timeIntervalSince(startedAt)
            
            writeVideoFrameAtTime(
                time: CMTime(value: CMTimeValue(millisElapsed),
                             timescale: 1000), image: image)
        }
        
        let processingSeconds = Date().timeIntervalSince(start)
        let delayRemaining = TimeInterval(interval) - processingSeconds
        
        
        let delay = delayRemaining > 0 ? delayRemaining : TimeInterval(interval)
        runRecordFrameLoop(delay: delay)
    }
    
    func completeRecordingSession() {
        videoWriterInput?.markAsFinished()
        
        var status = videoWriter?.status
        while status == .unknown {
            Thread.sleep(forTimeInterval: 0.5)
            status = videoWriter?.status
        }
        
        
        videoWriter?.finishWriting { [weak self] in
            let status = self?.videoWriter?.status
            
            if self?.videoWriter?.error != nil {
                let error = self?.videoWriter?.error
                DispatchQueue.main.async { [weak self] in
                    self?.callback?(nil, error)
                    self?.callback = nil
                }
            } else if status == .completed {
                DispatchQueue.main.async { [weak self] in
                    self?.callback?(self?.outputFile, nil)
                    self?.callback = nil
                }
            }
        }
        cleanupWriter()
    }
    
    
    func startRecordingUntil(_ timeLimitScs: TimeInterval, callback: @escaping StopRecordingCallBack) -> Bool {
        if recording {
            return false
        }
        
        initOrientation = UIApplication.shared.statusBarOrientation
        initScreenSize = UIScreen.main.bounds.size
        initScreenScale = fixedImageScale
        
        if !self.setupWriter() {
            return false
        }
        
        self.callback = callback
        startedAt = Date()
        timer = Timer.scheduledTimer(timeInterval: timeLimitScs, target: self, selector: #selector(timerFired(timer:)), userInfo: nil, repeats: false)
        recording = true
        runRecordFrameLoop(delay: 0)
        return true
    }
    
    func stopRecording() {
        if recording {
            timer?.invalidate()
            timer = nil
            recording = false
        }
    }
    
    func timerFired(timer: Timer) {
        stopRecording()
    }
    
    func writeVideoFrameAtTime(time: CMTime, image: UIImage) {
        guard let videoWriterInput = videoWriterInput, videoWriterInput.isReadyForMoreMediaData else {
            return
        }
        
        guard let pixelBufferPool = avAdaptor?.pixelBufferPool else {
            return
        }
        
        var pixelBuffer: CVPixelBuffer?
        let status = CVPixelBufferPoolCreatePixelBuffer(
            kCFAllocatorDefault,
            pixelBufferPool,
            &pixelBuffer)
        if status != kCVReturnSuccess {
            return
        }
        
        guard let cgImage = image.cgImage?.copy(), let data = cgImage.dataProvider?.data else {
            return
        }
        
        if let pixelBuffer = pixelBuffer,
            let destPixels = CVPixelBufferGetBaseAddress(pixelBuffer) {
            CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
            let range = CFRange(location: 0, length: CFDataGetLength(data))
            
            CFDataGetBytes(data, range, UnsafeMutablePointer<UInt8>(OpaquePointer(destPixels)))
            
            avAdaptor?.append(pixelBuffer, withPresentationTime: time)
            CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly)
        }
    }
}
